### PR TITLE
add  sigmoid focal loss operator for supporting retinanet

### PR DESCRIPTION
**Pull request by FlyingQianMM**
test=develop

Focal loss is used to address the foreground-background class imbalance existed on the training phase of one-stage detectors. This operator computes the sigmoid value for each element in the input tensor, after which focal loss is measured.
image

### CODE REVIEW

**Review by heavengate — APPROVED** [review]

**Review by heavengate — DISMISSED** [review]

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,210 @@
+/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc]: 2018 -> 2019

另外，这个op除过检测之外是不是个通用op

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu [hunk 2]
```diff
@@ -0,0 +1,178 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu]: 2016 -> 2019

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  return std::min((N + kNumCUDAThreads - 1) / kNumCUDAThreads,
+                  kNumMaxinumNumBlocks);
+}
+
+#define CUDA_1D_KERNEL_LOOP(i, n)                              \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+       i += blockDim.x * gridDim.x)
+
+template <typename T>
+__global__ void GPUSigmoidForward(const T *x_data, const int *label_data,
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu]: 这个不知Sigmoid功能吧

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.h [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      // p = 1. / 1. + expf(-x)
+      float p = 1. / (1. + std::exp(-x));
+
+      // (1 - p)**gamma * log(p) where
+      float term_pos =
+          std::pow((1. - p), gamma) * std::log(p > FLT_MIN ? p : FLT_MIN);
+      // p**gamma * log(1 - p)
+      float term_neg =
+          std::pow(p, gamma) *
+          (-1. * x * (x >= 0) - std::log(1. + std::exp(x - 2. * x * (x >= 0))));
```
Comment (qingqing01) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.h:67]: Why not T instead of float?

File: python/paddle/fluid/layers/detection.py [hunk 5]
```diff
@@ -210,6 +211,77 @@ def rpn_target_assign(bbox_pred,
     return predicted_cls_logits, predicted_bbox_pred, target_label, target_bbox, bbox_inside_weight
 
 
+def sigmoid_focal_loss(x, label, fg_num, gamma=2, alpha=0.25, num_classes=80):
```
Comment (qingqing01) [inline python/paddle/fluid/layers/detection.py]: 1. x的shape里已经有num class信息了，就不需要再让用户设置 num_classes了吧
2. fg_num始终是固定的吗？

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc (hunk 1 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc]: 是个通用op，不过目前应该只是在检测里会用到

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu (hunk 2 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu]: 已修改

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu (hunk 3 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/sigmoid_focal_loss_op.cu]: 还有softmax功能，不过标准的retinanet用的是sigmoid，所以暂时没有开发softmax focal loss

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.h (hunk 4 above)
Comment (FlyingQianMM) [reply paddle/fluid/operators/detection/sigmoid_focal_loss_op.h:67]: 已修改

File: python/paddle/fluid/layers/detection.py (hunk 5 above)
Comment (FlyingQianMM) [reply python/paddle/fluid/layers/detection.py]: 已去除num_classes，fg_num在每个batch中都是不一样的

**Review by qingqing01 — DISMISSED** [review]

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+using framework::Tensor;
+
+class SigmoidFocalLossOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("Label"), "Input(Label) should be not null.");
```
Comment (xsrobin) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc]: should not be

File: paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  }
+};
+
+class SigmoidFocalLossGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    PADDLE_ENFORCE(ctx->HasInput("X"), "Input(X) should be not null.");
+    PADDLE_ENFORCE(ctx->HasInput("Label"), "Input(Label) should be not null.");
```
Comment (xsrobin) [inline paddle/fluid/operators/detection/sigmoid_focal_loss_op.cc]: should not be

**Review by xsrobin — COMMENTED** [review]
as comments

**Review by qingqing01 — DISMISSED** [review]

**Review by xsrobin — DISMISSED** [review]
LGTM

**Review by qingqing01 — APPROVED** [review]

**Review by xsrobin — APPROVED** [review]
LGTM
